### PR TITLE
Hytera repeater support

### PR DESCRIPTION
**Pull request by zarya**
Support for the Hytera DMR repeaters

### CODE REVIEW

**Comment by crcro** [comment]
@zarya some code indentation cleanup is required:

if (expression) {

}

for (expression) {

}

**Comment by zarya** [comment]
Fixed

File: includes/discovery/os/hytera.inc.php [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+if (!$os || $os == "hytera") {
```
Comment (laf) [inline includes/discovery/os/hytera.inc.php]: Just needs to be `if (!os) {`

File: includes/discovery/os/hytera.inc.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    if (preg_match("/^lwIP/", $sysDescr)) {
```
Comment (laf) [inline includes/discovery/os/hytera.inc.php]: Can you not use the sysObjectId instead?

File: includes/discovery/sensors/temperatures/hytera.inc.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+if ($device['os'] == "hytera") {
+    $oids = snmp_walk($device, "rptPaTemprature", "-OsqnU","HYTERA-REPEATER-MIB");
+    if ($debug) {
```
Comment (laf) [inline includes/discovery/sensors/temperatures/hytera.inc.php]: Replace `if ($debug)` and the following lines with just `d_echo($oids);`

File: includes/discovery/sensors/temperatures/hytera.inc.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+if ($device['os'] == "hytera") {
+    $oids = snmp_walk($device, "rptPaTemprature", "-OsqnU","HYTERA-REPEATER-MIB");
+    if ($debug) { 
+        echo($oids."\n");
+    }
+    if ($oids !== false) echo("HYTERA-REPEATER-MIB ");
```
Comment (laf) [inline includes/discovery/sensors/temperatures/hytera.inc.php]: Needs to be full `if () {`

File: includes/discovery/sensors/voltages/hytera.inc.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+if ($device['os'] == "hytera") {
+    $oids = snmp_walk($device, "rptVoltage", "-OsqnU","HYTERA-REPEATER-MIB");
+    if ($debug) {
```
Comment (laf) [inline includes/discovery/sensors/voltages/hytera.inc.php]: Same as earlier

File: includes/discovery/sensors/voltages/hytera.inc.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        echo($oids."\n");
+    }
+    if ($oids !== false) echo("HYTERA-REPEATER-MIB ");
```
Comment (laf) [inline includes/discovery/sensors/voltages/hytera.inc.php]: Same as earlier.

File: includes/polling/functions.inc.php [hunk 7]
```diff
@@ -31,6 +31,9 @@ function poll_sensor($device, $class, $unit) {
                 else if ($device['os'] == 'canopy') {
                     include 'includes/polling/temperatures/canopy.inc.php';
                 }
+                else if ($device['os'] == 'hytera') {
```
Comment (laf) [inline includes/polling/functions.inc.php]: `} elseif () {`

File: includes/polling/functions.inc.php [hunk 8]
```diff
@@ -50,6 +53,22 @@ function poll_sensor($device, $class, $unit) {
                     }
                 }//end if
             }
+            else if ($class == "voltage") {
```
Comment (laf) [inline includes/polling/functions.inc.php]: `} elseif () {`

File: includes/polling/functions.inc.php [hunk 9]
```diff
@@ -50,6 +53,22 @@ function poll_sensor($device, $class, $unit) {
...
+                if ($device['os'] == 'hytera') {
+                    require_once "includes/polling/voltages/hytera.inc.php";
+                }
```
Comment (laf) [inline includes/polling/functions.inc.php]: `} else {`

File: includes/polling/functions.inc.php [hunk 10]
```diff
@@ -50,6 +53,22 @@ function poll_sensor($device, $class, $unit) {
...
+                else {
+                    for ($i = 0;$i < 5;$i++) {
```
Comment (laf) [inline includes/polling/functions.inc.php]: Where has the code this } else { come from? It voltage classes would have just fallen back to line 94.

Probably better to do:

`} elseif ($class == "voltage" && $device['os'] == 'hytera') {`

And just run the voltage include and then leave all other os' untouched.

**Comment by laf** [comment]
Some inline comments made + you need to look at the scrutinizer output: https://scrutinizer-ci.com/g/librenms/librenms/inspections/1fb91573-d70f-4f99-b7f5-e477db3235ef/issues/

File: includes/discovery/os/hytera.inc.php (hunk 2 above)
Comment (murrant) [inline includes/discovery/os/hytera.inc.php]: Yeah, this preg_match seems to generic and could have false positives.

File: includes/functions.php [hunk 11]
```diff
@@ -1463,3 +1463,73 @@ function report_this($message) {
     return '<h2>'.$message.' Please <a href="'.$config['project_issues'].'">report this</a> to the '.$config['project_name'].' developers.</h2>';
 
 }//end report_this()
+
+function hytera_h2f($number,$nd)
```
Comment (murrant) [inline includes/functions.php]: Does this function just change hex to float?

Could you just use floatval($hex) or (float)$hex?

Comment (zarya) [inline includes/functions.php]: floatval("90 4E 63 41") : 90
(float)"90 4E 63 41" : 90
hytera_h2f("90 4E 63 41",2) : 14.21

Comment (murrant) [inline includes/functions.php]: That's some special encoding right there :)

**Pull request merged**